src/components/forms/CharacterCount/CharacterCount.tsx

```diff
@@ -0,0 +1,165 @@
+    | null
+    | undefined
+}
+
+export const CharacterCount = (
+  props: CharacterCountProps &
+    React.InputHTMLAttributes<HTMLInputElement> &
+    React.TextareaHTMLAttributes<HTMLTextAreaElement>
+): React.ReactElement => {
+  const {
+    id,
+    name,
+    className,
+    label,
+    maxLength,
+    defaultValue,
+    isTextArea,
+    textRef,
+    type, // Conflicts with type attribute already set on TextInput component
+    ...inputProps
+  } = props
+
+  const defaultLength =
+    defaultValue === undefined ? 0 : defaultValue.toString().length
+  const maxNum = maxLength === undefined ? 0 : maxLength
+
+  /* Ideally defined as i18n translation strings */
+  const emptyMessageFormat = `${maxLength} characters allowed`
+  const remainingPluralFormat = '$0 characters left'
+  const remainingSingularFormat = '$0 character left'
+  const overSingularFormat = '$0 character over limit'
+  const overPluralFormat = '$0 characters over limit'
+
+  const limitMessage = (len: number): string => {
+    let limitMessage
+    if (len === 0) {
+      return emptyMessageFormat
+    } else if (len <= maxNum) {
+      switch (maxNum - len) {
+        case 1:
+          return remainingSingularFormat.replace('$0', '1')
+        default:
+          return remainingPluralFormat.replace('$0', (maxNum - len).toString())
+      }
+    } else {
+      switch (len - maxNum) {
+        case 1:
+          return overSingularFormat.replace('$0', '1')
+        default:
+          return overPluralFormat.replace('$0', (len - maxNum).toString())
+      }
+    }
+  }
+
+  const [state, setState] = useState({
+    characterLength: defaultLength,
+    limitMessage: limitMessage(defaultLength),
+    invalidMessage: defaultLength > maxNum,
+  })
+
+  const classes = classnames('usa-character-count__field', className)
+  const messageClasses = classnames(
+    'usa-hint',
+    'usa-character-count__message',
+    { 'usa-character-count__message--invalid': state.invalidMessage }
+  )
+
+  const messageId = id + '-info'
+
+  const handleChange = (
+    e: React.ChangeEvent<HTMLInputElement> &
+      React.ChangeEvent<HTMLTextAreaElement>
+  ): void => {
+    const len = e.target.value.length
+    setState({
+      characterLength: len,
+      limitMessage: limitMessage(len),
+      invalidMessage: len > maxLength,
+    })
+  }
+
+  const handleBlur = (
+    e: React.FocusEvent<HTMLInputElement> &
+      React.FocusEvent<HTMLTextAreaElement>
+  ): void => {
+    const validationMessage = state.invalidMessage
+      ? 'The content is too long.'
+      : ''
+    e.target.setCustomValidity(validationMessage)
+  }
+
+  let inputComponent
+  if (isTextArea) {
+    const ref = textRef as TextareaRef
+    // How will we know if the label or hint has an id value should we add it to the describedby?
+    inputComponent = (
+      <Textarea
+        id={id}
+        data-testid="characterCountInput"
+        name={name}
+        className={classes}
```

Comment:
the maxLength attribute should be added to each of the input component types: 
`maxLength={maxLength}`

Reply:
See https://designsystem.digital.gov/components/form-controls/#character-count in the implementation section.

Reply:
If we do that then the field will not allow input over `maxlength` characters.  Their javascript code reads that attribute, removes it then adds it as the `data-maxlength` to the character-count container in the uswds code.